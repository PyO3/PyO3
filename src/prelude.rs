//! PyO3's prelude.
//!
//! The purpose of this module is to alleviate imports of many commonly used items of the PyO3 crate
//! by adding a glob import to the top of pyo3 heavy modules:
//!
//! ```
//! # #![allow(unused_imports)]
//! use pyo3::prelude::*;
//! ```

pub use crate::conversion::{FromPyObject, IntoPyObject};
#[allow(deprecated)]
pub use crate::conversion::{IntoPy, ToPyObject};
pub use crate::err::{PyErr, PyResult};
pub use crate::instance::{Borrowed, Bound, Py, PyObject};
pub use crate::marker::Python;
pub use crate::pycell::{PyRef, PyRefMut};
pub use crate::pyclass_init::PyClassInitializer;
pub use crate::types::{PyAny, PyModule};

#[cfg(feature = "macros")]
pub use pyo3_macros::{
    pyclass, pyfunction, pymethods, pymodule, FromPyObject, IntoPyObject, IntoPyObjectRef,
};

#[cfg(feature = "macros")]
pub use crate::wrap_pyfunction;
#[cfg(feature = "macros")]
#[allow(deprecated)]
pub use crate::wrap_pyfunction_bound;

pub use crate::types::any::PyAnyMethods;
pub use crate::types::weakref::PyWeakrefMethods;
