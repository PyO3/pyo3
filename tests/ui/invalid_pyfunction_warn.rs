use pyo3::prelude::*;

#[pyfunction]
#[pyo3(warn)]
fn no_parenthesis_deprecated() {}

#[pyfunction]
#[pyo3(warn())]
fn no_message_deprecated() {}

#[pyfunction]
#[pyo3(warn(category = pyo3::exceptions::PyDeprecationWarning))]
fn no_message_deprecated_with_category() {}

#[pyfunction]
#[pyo3(warn(category = pyo3::exceptions::PyDeprecationWarning, message = ,))]
fn empty_message_deprecated_with_category() {}

#[pyfunction]
#[pyo3(warn(message = "deprecated function", category = ,))]
fn empty_category_deprecated_with_message() {}

#[pyfunction]
#[pyo3(warn(message = "deprecated function", random_key))]
fn random_key_deprecated() {}

#[pyclass]
struct DeprecatedMethodContainer {}

#[pymethods]
impl DeprecatedMethodContainer {
    #[classattr]
    #[pyo3(warn(message = "deprecated class attr"))]
    fn deprecated_class_attr() -> i32 {
        5
    }
}

#[pymethods]
impl DeprecatedMethodContainer {
    #[pyo3(warn(message = "deprecated __traverse__"))]
    fn __traverse__(&self, _visit: pyo3::gc::PyVisit<'_>) -> Result<(), pyo3::PyTraverseError> {
        Ok(())
    }
}

fn main() {}
