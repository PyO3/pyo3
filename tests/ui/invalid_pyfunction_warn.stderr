error: unexpected end of input, expected parentheses
 --> tests/ui/invalid_pyfunction_warn.rs:4:12
  |
4 | #[pyo3(warn)]
  |            ^

error: missing `message` in `warn` attribute
 --> tests/ui/invalid_pyfunction_warn.rs:8:13
  |
8 | #[pyo3(warn())]
  |             ^

error: missing `message` in `warn` attribute
  --> tests/ui/invalid_pyfunction_warn.rs:12:62
   |
12 | #[pyo3(warn(category = pyo3::exceptions::PyDeprecationWarning))]
   |                                                              ^

error: expected string literal
  --> tests/ui/invalid_pyfunction_warn.rs:16:74
   |
16 | #[pyo3(warn(category = pyo3::exceptions::PyDeprecationWarning, message = ,))]
   |                                                                          ^

error: expected identifier
  --> tests/ui/invalid_pyfunction_warn.rs:20:57
   |
20 | #[pyo3(warn(message = "deprecated function", category = ,))]
   |                                                         ^

error: expected `message` or `category`
  --> tests/ui/invalid_pyfunction_warn.rs:24:46
   |
24 | #[pyo3(warn(message = "deprecated function", random_key))]
   |                                              ^^^^^^^^^^

error: #[classattr] cannot be used with #[pyo3(warn)]
  --> tests/ui/invalid_pyfunction_warn.rs:33:12
   |
33 |     #[pyo3(warn(message = "deprecated class attr"))]
   |            ^^^^

error: __traverse__ cannot be used with #[pyo3(warn)]
  --> tests/ui/invalid_pyfunction_warn.rs:41:12
   |
41 |     #[pyo3(warn(message = "deprecated __traverse__"))]
   |            ^^^^
